### PR TITLE
feat: Implement Drawer Navigation

### LINKED ISSUE #148: Make event container page starter page after log in
Currently after log in event list page is a starter page but for event organizer, starter page should be event container page for a selected event. And from navigation drawer organizer can go to the event list page and select the required event. 

@niranjan94 @iamareebjamal what is your view ?

### PR DESCRIPTION
**Pull request by mejariamol**
- Replaces EventListActivity with EventListFragment
- Adds MainAcivity which hosts all the fragments with Navigation Drawer Support

fixes #148 

screenshot_2017-07-03-04-12-00

### CODE REVIEW

**Comment by mejariamol** [comment]
@iamareebjamal @rohanagarwal94 @Shailesh351 @arp95 Please review.

File: app/src/main/java/org/fossasia/openevent/app/data/UtilModel.java [hunk 1]
```diff
@@ -60,6 +60,18 @@ public void saveString(String key, String value) {
     }
 
     @Override
+    public long getLong(String key, long defaultValue) {
+        return sharedPreferences.getLong(key, -1);
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/UtilModel.java]: Use defaultValue as second argument

File: app/src/main/java/org/fossasia/openevent/app/events/EventsListAdapter.java [hunk 2]
```diff
@@ -141,24 +115,14 @@ public int getItemCount() {
...
-                if (isTwoPane) {
-                    showEvent(event);
-                } else {
-                    intent.putExtra(EventListActivity.EVENT_KEY, event.getId());
-                    intent.putExtra(EventListActivity.EVENT_NAME, event.getName());
-                    context.startActivity(intent);
-                }
-            });
+
+            binding.getRoot().setOnClickListener(view -> ((OnEventLoadedListener)context).onEventLoaded(event));
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/events/EventsListAdapter.java]: Why is this needed? Custom listeners are bad

Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/events/EventsListAdapter.java]: I thought it is a good way to pass data from fragment to activity.

File: app/src/main/java/org/fossasia/openevent/app/data/UtilModel.java (hunk 1 above)
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/data/UtilModel.java]: Yeah. Thats mistake. :+1:

**Comment by mejariamol** [comment]
>A lot of changes in MainActivity

Yeah, fragments handling logic need to be improved. Please do suggest. @iamareebjamal

**Comment by mejariamol** [comment]
@iamareebjamal waiting for your suggestions on _MainActivity_.

**Comment by iamareebjamal** [comment]
Yes @mejariamol Was busy with 3 PRs yesterday. Will surely take a look today

**Comment by iamareebjamal** [comment]
@mejariamol There is some binding error trying it locally. Can you please take a look?

**Comment by iamareebjamal** [comment]
1. There's a memory leak in AttendeeFragment
2. The method of updating attendee is convoluted again. Getting item position of the attendee and updating it is bad
3. Events -> Dashboard and other interactions are flat. Meaning that if you press back, it exits the app, whereas user expects that it will take me back to events page. I'll take a look at EventBrite app and suggest what's the better approach, but current one is confusing
4. Keep the navigation header light grey and empty for now as the current info isn't completely shown and the PR is already huge
5. Don't use dividers in Navigation Drawer. It's non standard and use icons and support selected menu item o that user is aware on which section he is


These are just some UI related points I recommend. As this is complete UI change, this may take long to get right.

**Comment by mejariamol** [comment]
Working on it.
- [ ] There's a memory leak in AttendeeFragment
- [ ] The method of updating attendee is convoluted again. Getting item position of the attendee and updating it is bad
- [ ] Events -> Dashboard and other interactions are flat. Meaning that if you press back, it exits the app, whereas user expects that it will take me back to events page.
- [ ] Keep the navigation header light grey and empty for now as the current info isn't completely shown and the PR is already huge
- [ ] Don't use dividers in Navigation Drawer. It's non standard and use icons and support selected menu item o that user is aware on which section he is
- [ ] Log Out should be in another section of navigation drawer

**Comment by mejariamol** [comment]
@iamareebjamal 
I haven't used [NavigationView](https://developer.android.com/reference/android/support/design/widget/NavigationView.html) (provided by support library) for side drawer, thinking that we can customize it as per our need. Is it fine or should we use support library's NavigationView? I am not sure if it enables adding menu items dynamically.

**Comment by mejariamol** [comment]
@iamareebjamal

screenshot_2017-07-02-03-06-12

**Comment by iamareebjamal** [comment]
Looks perfect

**Comment by iamareebjamal** [comment]
About the data. I think NavigationView supports it. I'll confirm that

**Comment by mejariamol** [comment]
I have used ListView for navigation drawer menus.

**Comment by iamareebjamal** [comment]
Ok, let's go with ListView only, but can we use DataBinding? I saw that you were previously manually loading stuff

**Comment by mejariamol** [comment]
@iamareebjamal please review

File: app/src/main/java/org/fossasia/openevent/app/event/detail/EventDetailFragment.java [hunk 3]
```diff
@@ -100,9 +103,21 @@ public void onActivityCreated(@Nullable Bundle savedInstanceState) {
     }
 
     @Override
+    public void onAttach(Context context) {
+        super.onAttach(context);
+        if (context instanceof OnEventLoadedListener) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/event/detail/EventDetailFragment.java]: This is not the correct way. Interfaces were made so that you don't cast classes. However, as I think we'll be changing this pattern, I'll think of other solution for communication between fragments

File: app/src/test/java/org/fossasia/openevent/app/presenter/MainPresenterTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import io.reactivex.Completable;
+import io.reactivex.observers.TestObserver;
+import io.reactivex.plugins.RxJavaPlugins;
+import io.reactivex.schedulers.Schedulers;
+
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(JUnit4.class)
+public class MainPresenterTest {
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/presenter/MainPresenterTest.java]: Leave tests for now as PR is getting huge already

File: app/src/main/res/layout/fragment_event_list.xml [hunk 5]
```diff
@@ -0,0 +1,38 @@
+<android.support.design.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (iamareebjamal) [inline app/src/main/res/layout/fragment_event_list.xml]: This is exactly similar to the `event_list` xml except that it doesn't have data binding, which is going back in history as we will definitely use databinding in future to show and hide empty layout, so please don't edit it and it can be reused as is. This will also reduce the changes in repository

File: app/src/main/res/layout/event_list_container.xml [hunk 6]
```diff
@@ -1,23 +0,0 @@
-<layout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (iamareebjamal) [inline app/src/main/res/layout/event_list_container.xml]: Does this mean we'll lose the tablet layout?

File: app/src/main/res/layout/drawer_list_item.xml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            android:layout_height="@dimen/spacing_normal"
+            android:gravity="center_vertical"
+            android:visibility="@{menu.belowSeparator ? View.VISIBLE : View.GONE}">
+
+            <RelativeLayout
+                android:layout_width="match_parent"
+                android:layout_height="1dp"
+                android:background="#e4e4e4"/>
+
+        </LinearLayout>
```
Comment (iamareebjamal) [inline app/src/main/res/layout/drawer_list_item.xml]: This layout is too complex for a divider. Use FrameLayout?

File: app/src/main/res/layout/drawer_list_item.xml [hunk 8]
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout xmlns:android="http://schemas.android.com/apk/res/android">
+    <data>
+        <import type="android.view.View"/>
+        <variable
+            name="menu"
+            type="org.fossasia.openevent.app.main.menu.OpevMenu"/>
```
Comment (iamareebjamal) [inline app/src/main/res/layout/drawer_list_item.xml]: Are we using a new class just for menu? I think these complexities can be heavily reduced if we follow standard material guidelines for creating NavView

File: app/src/main/res/layout/drawer_layout.xml [hunk 9]
```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.v4.widget.DrawerLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@+id/drawerLayout"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <include layout="@layout/activity_main" />
```
Comment (iamareebjamal) [inline app/src/main/res/layout/drawer_layout.xml]: This should be opposite. MainActivity should include drawer, not the other way around

File: app/src/main/res/layout/content_main.xml [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/fragmentContainer"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    app:layout_behavior="@string/appbar_scrolling_view_behavior"
+    tools:context="org.fossasia.openevent.app.main.MainActivity"
+    tools:showIn="@layout/activity_main">
+
+</FrameLayout>
```
Comment (iamareebjamal) [inline app/src/main/res/layout/content_main.xml]: This is empty FrameLayout?

File: app/src/main/res/layout/event_list_container.xml (hunk 6 above)
Comment (mejariamol) [inline app/src/main/res/layout/event_list_container.xml]: Yes,  tablet implementation will need to re-implement.

File: app/src/main/res/layout/drawer_list_item.xml (hunk 8 above)
Comment (mejariamol) [inline app/src/main/res/layout/drawer_list_item.xml]: It will be beneficial if we make menu class. So that in future we can add attributes to menu item. For example attendeee number or check in checked out number.

Comment (iamareebjamal) [inline app/src/main/res/layout/drawer_list_item.xml]: But let's take it one step at a time. Use menu xml for now.

Comment (mejariamol) [inline app/src/main/res/layout/drawer_list_item.xml]: You mean `android.support.design.widget.NavigationView`  over `ListView` ?

File: app/src/main/res/layout/drawer_list_item.xml (hunk 7 above)
Comment (mejariamol) [inline app/src/main/res/layout/drawer_list_item.xml]: actually section divider has to be 16dp thick with line at middle. Thats why this. FrameLayout wil just replace LinearLayout here.

File: app/src/main/res/layout/drawer_layout.xml (hunk 9 above)
Comment (mejariamol) [inline app/src/main/res/layout/drawer_layout.xml]: doing it. :+1:

File: app/src/main/res/layout/content_main.xml (hunk 10 above)
Comment (mejariamol) [inline app/src/main/res/layout/content_main.xml]: fragment host it is. Will make it `/>`

File: app/src/main/res/layout/drawer_list_item.xml (hunk 8 above)
Comment (iamareebjamal) [inline app/src/main/res/layout/drawer_list_item.xml]: Yes, I'm talking about NavgationView for now

File: app/src/main/res/layout/drawer_list_item.xml (hunk 7 above)
Comment (iamareebjamal) [inline app/src/main/res/layout/drawer_list_item.xml]: LinearLayout with RelativeLayout is too complex. FrameLayout with padding can do the same thiing. Nevertheless, we should use NavigationView for now

File: app/src/main/res/layout/content_main.xml (hunk 10 above)
Comment (iamareebjamal) [inline app/src/main/res/layout/content_main.xml]: If it is this simple, we can add it in the layout itself, rather host it here

File: app/src/main/res/layout/drawer_list_item.xml (hunk 7 above)
Comment (mejariamol) [inline app/src/main/res/layout/drawer_list_item.xml]: Okay. Using navigation view for now. :+1:

**Comment by mejariamol** [comment]
@iamareebjamal please review. have tried to keep things simpler.

File: app/src/main/java/org/fossasia/openevent/app/main/MainActivity.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+        if (id == R.id.nav_logout) {
+            presenter.logout();
+        } else {
+            if (eventId != -1) {
+                loadFragment(item.getItemId());
+            }
+        }
+
+        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/main/MainActivity.java]: Use butterknife

File: app/src/main/java/org/fossasia/openevent/app/main/MainActivity.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        }
+        backPressed = System.currentTimeMillis();
+    }
+
+    @SuppressWarnings("StatementWithEmptyBody")
+    @Override
+    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
+        int id = item.getItemId();
+
+        if (id == R.id.nav_logout) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/main/MainActivity.java]: Use switch case

File: app/src/main/java/org/fossasia/openevent/app/main/MainActivity.java [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+                fragment = AttendeesFragment.newInstance(eventId);
+                break;
+            case R.id.nav_events:
+                fragment = EventListFragment.newInstance();
+                break;
+            default:
+                fragment = EventDetailFragment.newInstance(eventId);
+        }
+        setTitle(navigationView.getMenu().findItem(navItemId).getTitle());
+        FragmentManager fm = getSupportFragmentManager();
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/main/MainActivity.java]: Store this in a global variable so that it does not need to get queried again and again

File: app/src/main/java/org/fossasia/openevent/app/main/MainPresenter.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    private IMainView mainView;
+    private ILoginModel loginModel;
+
+    @Inject
+    public MainPresenter(ILoginModel loginModel) {
+        this.loginModel = loginModel;
+    }
+
+    @Override
+    public void attach(IMainView mainView) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/main/MainPresenter.java]: Add detach too

File: app/src/main/res/values/colors.xml [hunk 15]
```diff
@@ -3,6 +3,7 @@
     <color name="color_primary">#EEEEEE</color>
     <color name="color_primary_dark">#1976d2</color>
     <color name="color_accent">#2196f3</color>
+    <color name="color_accent_light">#79c0f7</color>
```
Comment (iamareebjamal) [inline app/src/main/res/values/colors.xml]: Rather name the color like blue a400

File: app/src/main/java/org/fossasia/openevent/app/events/EventListFragment.java [hunk 16]
```diff
@@ -1,162 +1,138 @@
...
-    public static final String EVENT_NAME = "event_name";
+public class EventListFragment extends BaseFragment implements IEventsView{
+    @BindView(R.id.swipeContainer)
+    SwipeRefreshLayout refreshLayout;
+    @BindView(R.id.eventRecyclerView)
+    RecyclerView recyclerView;
+    @BindView(R.id.progressBar)
+    ProgressBar progressBar;
+    @BindView(R.id.emptyView)
+    View emptyView;
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/events/EventListFragment.java]: Why not using data binding

Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/events/EventListFragment.java]: better we use butterknife for handling views and databinding for setting dynamic properties

File: app/src/main/java/org/fossasia/openevent/app/main/MainActivity.java (hunk 12 above)
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/main/MainActivity.java]: it is just an if-else statement

File: app/src/main/java/org/fossasia/openevent/app/main/MainPresenter.java (hunk 14 above)
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/main/MainPresenter.java]: done :+1:

File: app/src/main/java/org/fossasia/openevent/app/events/EventListFragment.java (hunk 16 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/events/EventListFragment.java]: We'll have to change the layout again to data binding format when we move transient logic in. We should use data binding wherever possible

Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/events/EventListFragment.java]: yeah. But using databinding for view inflation is not helping in code readability. It is just same as using `findViewById`. We must use databinding for setting properties. but I think we should use butterknife for view inflation.  what do you say? @iamareebjamal

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/events/EventListFragment.java]: The point is to not cause changes which are to be rechanged again in future, plus, databinding is more efficient and we'll have to shift to it in near future for transient field, so why change a layout already in data binding back to a simple one? If it would be already not in data binding, I wouldn't be suggesting it to be made that way. This is like undoing something only to redo later, involving more work both now and then

**Comment by mejariamol** [comment]
@iamareebjamal have used databinding for view inflation in event list fragment. Please review. :)

**Comment by iamareebjamal** [comment]
OK, let's merge it and then solve the complexities after. But first, let's wait for the simpler PRs to get merged. Sounds good?

**Comment by niranjan94** [comment]
@mejariamol please resolve conflicts

**Comment by mejariamol** [comment]
@iamareebjamal please review.

**Comment by mejariamol** [comment]
@iamareebjamal please review :smile:

**Comment by mejariamol** [comment]
@niranjan94 please merge
